### PR TITLE
PHP to Python plus tests and stuff

### PR DESCRIPTION
**Pull request by fabianegli**
This PR changes the behaviour:

- ports php code to python -> no more PHP code, one language left
- adds test for the json to txt and md conversion
- removes the action for the php script - everything is now updated at once with the Python script
- adds the action trigger push to the remaining action
- adds some Python specific files and folders to .gitignore

and some file contents slightly:

- removes the space at end of some lines in md table
- adds a newline at end of robots.txt file

The python files have been formatted with black.

### CODE REVIEW

File: .github/workflows/daily_update.yml [hunk 1]
```diff
@@ -1,5 +1,8 @@
 name: Daily Update from Dark Visitors
 on:
+  push:
```
Comment (glyn) [inline .github/workflows/daily_update.yml:3]: I think this file needs renaming now that it is more than just daily updates.

There could be a small downside to combining the function in this way: updates on merge could fail if the dark visitors site is down. But I like doing the dark visitors merge in more cases, e.g. if we change the dark visitors handling code, we'd want to re-run it on merge. So, on balance, it's probably a good approach.

File: code/tests.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+"""These tests can be run with pytest.
```
Comment (glyn) [inline code/tests.py:1]: It's wonderful to see tests being added. That said, I wonder if now is the time to test the dark visitors loading and merging code too?

File: code/dark_visitors.py [hunk 3]
```diff
@@ -4,72 +4,103 @@
...
-            "description": consolidate("description", f"{desc} More info can be found at https://darkvisitors.com/agents{agent['href']}")
-        }
-
-print(f"Total: {len(existing_content)}")
-sorted_keys = sorted(existing_content, key=lambda k: k.lower())
-existing_content = {k: existing_content[k] for k in sorted_keys}
-Path("./robots.json").write_text(json.dumps(existing_content, indent=4))
\ No newline at end of file
+
+def get_updated_robots_json():
```
Comment (glyn) [inline code/dark_visitors.py:8]: It would be great to include the bulk of this logic in the tests, which I think would require factoring out a pure function or similar. See my comment about testing later.

File: .github/workflows/daily_update.yml [hunk 4]
```diff
@@ -22,9 +25,3 @@ jobs:
           git add -A
           git diff --quiet && git diff --staged --quiet || (git commit -m "Daily update from Dark Visitors" && git push)
```
Comment (glyn) [inline .github/workflows/daily_update.yml:26]: The above commit message is not appropriate for a merge of a PR. I wonder if we need to keep the actions separate and use different commit messages in each case?

File: code/test_files/robots.json [hunk 5]
```diff
@@ -0,0 +1,282 @@
+{
```
Comment (glyn) [inline code/test_files/robots.json:1]: I wonder if it's worth reworking this file so that it has less duplication of fields and so that it exercises the consolidation logic thoroughly if testing of dark visitors update is added?

**Review by glyn — CHANGES_REQUESTED** [review]
Thanks for this. I like the idea of centralising on a single coding language. We don't know if anyone will maintain the code, but at least we'll only need Python skills, which are more common than PHP skills. Also the addition of tests is really appreciated.

I've suggested various improvements. You may regard some of the larger ones as out of scope for this PR, in which case, please log an issue for each of the ones you choose not to address so these can be worked on in future. Alternatively, if you don't mind large PRs, please address all the comments here - your choice.

**Comment by glyn** [comment]
@cdransf PTAL as this is a substantial PR.

File: .github/workflows/daily_update.yml (hunk 1 above)
Comment (fabianegli) [reply .github/workflows/daily_update.yml:3]: A better name would be useful. What should be though? `ai_robots_update`?

I did not consider this, but will address this in this PR.

Comment (glyn) [reply .github/workflows/daily_update.yml:3]: Let's think about this. The inputs are `robots.json` and information from the dark visitors site and the outputs are `robots.txt` and `table-of-bot-metrics.md`. In other words, it's complicated. So I think some generic name like "generate" or "update" might be enough, given the context.

Comment (glyn) [reply .github/workflows/daily_update.yml:3]: Oh and the action name should use the same terminology as the filename, but can be a bit wordier.

Comment (fabianegli) [reply .github/workflows/daily_update.yml:3]: The actions are quite fast. So they can easily run on each push by my judgement.

I did add some error handling for the case the script is run and cannot connect to darkvisitors.com.

File: code/tests.py (hunk 2 above)
Comment (fabianegli) [reply code/tests.py:1]: It would be nice to have that, but this is outside my current purvey.

File: .github/workflows/daily_update.yml (hunk 4 above)
Comment (fabianegli) [reply .github/workflows/daily_update.yml:26]: I agree. This is a new inconsistency. I'll see what I can do.

File: code/test_files/robots.json (hunk 5 above)
Comment (fabianegli) [reply code/test_files/robots.json:1]: I am not entirely sure what you are telling me here. But changing file formats is not my intention here. Maybe elaborate more in a new issue.

File: code/dark_visitors.py (hunk 3 above)
Comment (fabianegli) [reply code/dark_visitors.py]: I don't have a clear idea how to do that sensibly. Maybe someone else has? But this is in my eyes no blocker for this PR.

**Comment by fabianegli** [comment]
Thanks for the review @glyn

File: .github/workflows/ai_robots_update.yml [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    runs-on: ubuntu-latest
+    name: dark-visitors
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 2
+      - run: |
+          pip install beautifulsoup4 requests
+          git config --global user.name "dark-visitors"
+          git config --global user.email "dark-visitors@users.noreply.github.com"
```
Comment (fabianegli) [inline .github/workflows/ai_robots_update.yml:20]: Why are these configs made here? And should they be changed for the conversion step?

Comment (glyn) [reply .github/workflows/ai_robots_update.yml:20]: I don't know (x2).

File: code/dark_visitors.py (hunk 3 above)
Comment (glyn) [reply code/dark_visitors.py]: I agree this can be done as a follow-on PR (not necessarily by you). We need to remember to raise an issue to get that task on the books.

Comment (glyn) [reply code/dark_visitors.py]: I raised https://github.com/ai-robots-txt/ai.robots.txt/issues/52 to cover this.

File: .github/workflows/daily_update.yml (hunk 4 above)
Comment (glyn) [reply .github/workflows/daily_update.yml]: Thanks.

File: .github/workflows/ai_robots_update.yml (hunk 6 above)
Comment (glyn) [reply .github/workflows/ai_robots_update.yml:20]: Actually, the purpose of these configs is to make the changes produced by the dark visitors website look as if they were produced by a dark-visitors user. I'm not sure this is entirely necessary. The commit message should be sufficient on its own.

**Comment by glyn** [comment]
@fabianegli Thanks for the rework. A couple of remaining items for this PR:

1. rationalise the git config user name/email so that updates are not wrongly attributed to dark-visitors. (I haven't worked through to know whether this has already been done, but please take a look.)
2. the tests need to be run each time an action is triggered. Have I missed where that happens?

**Comment by glyn** [comment]
@fabianegli To expand on remaining item 1: please make sure that updates due to dark visitors changes have an appropriate commit message (and user/email if you decide to keep that) and that updates due to merging a local change also have an appropriate commit message (and user/email). In other words, we need to make sure we differentiate between dark visitors and local changes, especially in the situation where a local change is merged and, coincidentally, dark visitors has changed. If we are not careful, the current code will pick up both changes and put them in a single commit and give that commit an incorrect commit message (and user/email).

It may be necessary to step back from the approach in the current PR and keep the processing of dark visitors separate from the merging of local changes. It should still be possible to (a) replace the PHP by Python and (b) introduce tests of the Python functions.

Hope that helps make the requirements clearer.

**Comment by newbold** [comment]
Just gonna “Leeroy Jenkins” this and we’ll see what happens!

**Comment by newbold** [comment]
It looks like the action ran without any issue. Not sure who can see this, but here’s the run that was triggered via the merge: https://github.com/ai-robots-txt/ai.robots.txt/actions/runs/11471373718

No changes to robots.json or robots.txt, though, but I can’t tell if that’s because there truly aren’t any new user agents to add or if something else happened.

**Comment by glyn** [comment]
> Just gonna “Leeroy Jenkins” this and we’ll see what happens!

Will all changes now be attributed to dark visitors?

**Comment by fabianegli** [comment]
I think there was indeed a change in the attribution logic. PR #58 reverts that.

**Comment by fabianegli** [comment]
I am sorry it took me sol long to come back to this and indeed this probably shouldn't have been merged in this way. FYI that's why I marked it to be a draft before I had to turn my attention to other things for a while.
